exchanges/gemini/gemini_websocket.go

```diff
@@ -52,179 +57,267 @@ func (g *Gemini) WsSubscribe(dialer *websocket.Dialer) error {
-
-		endpoint := fmt.Sprintf(g.Websocket.GetWebsocketURL(),
+		endpoint := fmt.Sprintf("%s%s/%s?%s",
+			g.WebsocketURL,
+			geminiWsMarketData,
 			c.String(),
 			val.Encode())
-
-		conn, _, err := dialer.Dial(endpoint, http.Header{})
+		conn, conStatus, err := dialer.Dial(endpoint, http.Header{})
 		if err != nil {
-			return err
+			return fmt.Errorf("%v %v %v Error: %v", endpoint, conStatus, conStatus.StatusCode, err)
 		}
-
-		go g.WsReadData(conn, c, geminiWsMarketData)
-
+		go g.WsReadData(conn, c)
 		if len(enabledCurrencies)-1 == i {
 			return nil
 		}
+	}
+	return nil
+}
 
-		time.Sleep(5 * time.Second) // rate limiter, limit of 12 requests per
-		// minute
+// WsSecureSubscribe will connect to Gemini's secure endpoint
+func (g *Gemini) WsSecureSubscribe(dialer *websocket.Dialer, url string) error {
+	if !g.GetAuthenticatedAPISupport(exchange.WebsocketAuthentication) {
+		return fmt.Errorf("%v AuthenticatedWebsocketAPISupport not enabled", g.Name)
 	}
+	payload := WsRequestPayload{
+		Request: fmt.Sprintf("/v1/%v", url),
+		Nonce:   time.Now().UnixNano(),
+	}
+	PayloadJSON, err := common.JSONEncode(payload)
+	if err != nil {
+		return fmt.Errorf("%v sendAuthenticatedHTTPRequest: Unable to JSON request", g.Name)
+	}
+
+	endpoint := fmt.Sprintf("%v%v", g.WebsocketURL, url)
+	PayloadBase64 := common.Base64Encode(PayloadJSON)
+	hmac := common.GetHMAC(common.HashSHA512_384, []byte(PayloadBase64), []byte(g.APISecret))
+	headers := http.Header{}
+	headers.Add("Content-Length", "0")
+	headers.Add("Content-Type", "text/plain")
+	headers.Add("X-GEMINI-PAYLOAD", PayloadBase64)
+	headers.Add("X-GEMINI-APIKEY", g.APIKey)
+	headers.Add("X-GEMINI-SIGNATURE", common.HexEncodeToString(hmac))
+	headers.Add("Cache-Control", "no-cache")
+
+	conn, conStatus, err := dialer.Dial(endpoint, headers)
+	if err != nil {
+		return fmt.Errorf("%v %v %v Error: %v", endpoint, conStatus, conStatus.StatusCode, err)
+	}
+	go g.WsReadData(conn, currency.Pair{})
 	return nil
 }
 
 // WsReadData reads from the websocket connection and returns the websocket
 // response
-func (g *Gemini) WsReadData(ws *websocket.Conn, c currency.Pair, feedType string) {
+func (g *Gemini) WsReadData(ws *websocket.Conn, c currency.Pair) {
 	g.Websocket.Wg.Add(1)
-
-	defer func() {
-		err := ws.Close()
-		if err != nil {
-			g.Websocket.DataHandler <- fmt.Errorf("gemini_websocket.go - Unable to to close Websocket connection. Error: %s",
-				err)
-		}
-		g.Websocket.Wg.Done()
-	}()
-
+	defer g.Websocket.Wg.Done()
 	for {
 		select {
 		case <-g.Websocket.ShutdownC:
 			return
-
 		default:
 			_, resp, err := ws.ReadMessage()
 			if err != nil {
 				g.Websocket.DataHandler <- err
 				return
 			}
-
 			g.Websocket.TrafficAlert <- struct{}{}
-			comms <- ReadData{Raw: resp, Currency: c, FeedType: feedType}
+			comms <- ReadData{Raw: resp, Currency: c}
 		}
 	}
-
 }
 
 // WsHandleData handles all the websocket data coming from the websocket
 // connection
 func (g *Gemini) WsHandleData() {
```

Comment:
@shazbert this comment should be on [line 93](https://github.com/thrasher-/gocryptotrader/pull/315/files#diff-69dcd26a324bd2cd00cdb0b260b9ada3R93), looks like there's a bug on our end. [Issue](https://github.com/codelingo/codelingo/issues/424).

[Effective Go opposes underscores](https://golang.org/doc/effective_go.html#mixed-caps), but `HashSHA512_384` seems like a perfectly reasonable exception since digits have no case. I've [edited the tenet](https://github.com/codelingo/codelingo/pull/425) to allow underscores separating numerical characters.

Please @ me or any of the CodeLingo team if you see issues like this 😄

Reply:
:+1: